change_date_rename_file.sh

```diff
@@ -1,31 +1,83 @@
-# & Modified Date to the photo-taken date.
-#
-# directory_path is then referred to as $1 - which refers to the 1st argument
-# passed from the command line.
+# Brief: Command-line Bash script to change Creation Date & Modified Date that  
+# are currently download date to photo-taken date. The script also adds
+# photo-taken date to the filename, e.g., 2015-09-02_07-09_0059.jpg.
+# These changes will make file sorting easier. Takes in 1 command-line 
+# parameter, directory_path, the location of the files.
 #
 # Note: Photo-taken date is exif:DateTimeOriginal.
 # Note: Script processes only a single directory. 
 #
-# Kim Lew
-
-test_path=/Users/kimlew/Documents/PHOTOS/2019/iPhone_photos_2019-Jan-01-Jul-30
+# Author: Kim Lew
 
 echo "Type the directory path for the files that need changed dates & filenames: "
-read dirname
-echo "You typed: " $dirname
-
-# while read filename; do
-identify -format '%[EXIF:DateTimeOriginal]' \
-$test_path/IMG_0059.jpg \
-| sed -e 's/.\{3\}$//' -e 's/:/-/g' -e 's/ /_/g' \
-> $test_path/date_manipulation/aDateTimeOriginal2.txt
-# done
+read directory_path
+echo "You typed: " $directory_path
+
+if [ ! -d $directory_path ] 
+then
+    echo "This directory does NOT exist." 
+    exit 9999 # die with error code 9999
+fi
+
+if ! which identify > /dev/null; then
+  echo "Error: You are missing the identify program that is part of the"
+  echo "ImageMagick software suite. Install ImageMagick with your package manager."
+  echo "Or see: https://imagemagick.org/index.php/"
+  echo "Then re-run this script."
+  exit 9999 # die with error code 9999
+fi
+
+echo "Date changes and filename changes in progress..."
+echo "..."
+
+# Loop that processes entire given directory.
+find $directory_path -type f -name '*.jpg' |
+while read a_file_name; do
+  exif_date=$(identify -format '%[EXIF:DateTimeOriginal]' $a_file_name)
+  
+  if [ "$exif_date" == '' ] > /dev/null; then
+    echo "Error: The file, $a_file_name"
+    echo "- is missed the exif metadata, EXIF:DateTimeOriginal - so skipping this file"
+    echo "Continuing with next file ..."
+    echo
+    continue
+  fi
+  ### Filesystem/OS Date Change ###
+  # Change filesystem date to EXIF photo-taken date, EXIF:DateTimeOriginal.
+  # 1. Replace ALL occurrences of : 	With: nothing
+  # 2. Replace 1st occurrence of space 	With: nothing
+  # 3. Build string by deleting last 2 char at end. Then concat . & then concat 
+  # last 2 char, e.g., abc12 -> abc + . + 12 => abc.12
+
+  date_for_date_change="${exif_date//:/}"
+  date_for_date_change="${date_for_date_change/ /}"  
+
+  # Use format: ${parameter%word} for the portion with the string to keep.
+  # % - means to delete only the following stated chars & keep the rest, i.e., 
+  # %${date_for_date_change: -2} - which is the 12 part of abc12 & keep abc
+  date_for_date_change="${date_for_date_change%${date_for_date_change: -2}}.${date_for_date_change: -2}"
```

Comment:
Suggestion: there is a way to shorten this last change in `date_for_data_change` that might make it a bit more readable. A simpler way to discard the last two characters of the value of `data_for_date_change` is:

```
${data_for_date_change%??}
```
Each question mark means 'any character', so this drops the last two characters. You then get

```shell
date_for_date_change="${date_for_date_change%??}.${date_for_date_change: -2}"
```

If you don't think that's easier to read then just ignore this suggestion.

Reply:
@kcharter That is WAY more readable. Thanks.